driver-core/src/main/java/com/datastax/driver/core/policies/ErrorAwarePolicy.java

```diff
@@ -0,0 +1,313 @@
+            this.errorsFilter = errorsFilter;
+            return this;
+        }
+
+        /**
+         * Defines the time during which a host is excluded by the tracker.
+         * <p/>
+         * When a host's error rate overpasses the threshold specified in {@link #withInclusionThreshold}, the host
+         * will be considered excluded, and this for a certain period of time defined by this method. After this period,
+         * a previously excluded host will be included again unless/until it reaches the threshold again.
+         * <p/>
+         * If the error tracker, during an update, detects that a currently excluded host continues to overpass the
+         * error rate threshold, this host's exclusion time will be reset to the start.
+         * <p/>
+         * Default value for the retry period is 5 seconds.
+         *
+         * @param retryPeriod the period to use for refreshing the ErrorTracker.
+         * @return this {@link Builder} instance, for method chaining.
+         */
+        public Builder withRetryPeriod(int retryPeriod) {
+            this.retryPeriod = retryPeriod;
+            return this;
+        }
+
+        /**
+         * Creates the {@link ErrorAwarePolicy} instance.
+         * @return the newly created {@link ErrorAwarePolicy}.
+         */
+        public ErrorAwarePolicy build() {
+            return new ErrorAwarePolicy(this);
+        }
+    }
+
+    class PerHostErrorTracker implements LatencyTracker {
+
+        private double errorsThreshold;
+
+        private ConcurrentMap<Host, Meter> hostsMeters = new ConcurrentHashMap<Host, Meter>();
+
+        private ConcurrentMap<Host, Long> exclusions = new ConcurrentHashMap<Host, Long>();
+
+        private ErrorsFilter errorsFilter;
+
+        PerHostErrorTracker(double errorsThreshold, ErrorsFilter errorsFilter) {
+            this.errorsThreshold = errorsThreshold;
+            this.errorsFilter = errorsFilter;
+        }
+
+        @Override
+        public void update(Host host, Statement statement, Exception exception, long newLatencyNanos) {
+            if (exception == null) {
+                return;
+            }
+            if (!errorsFilter.shouldAccountForError(exception, host)) {
+                return;
+            }
+            if (!hostsMeters.containsKey(host)) {
+                hostsMeters.putIfAbsent(host, new Meter());
+            }
+            Meter hostMeter = hostsMeters.get(host);
+            hostMeter.mark();
+            if (hostMeter.getOneMinuteRate() >= errorsThreshold) {
+                logger.debug(String.format("Host : %s has reached the maximum authorized number of errors. Its error rate is : %.3f. " +
+                                "It will now be ignored by the ErrorAware load balancing policy during a period of %d milliseconds.",
+                        host, hostMeter.getOneMinuteRate(), retryPeriodMillis));
+                exclusions.putIfAbsent(host, System.currentTimeMillis());
```

Comment:
True `putIfAbsent` isn't good if we don't clear, however I'd rather replace it with `put`, rather than clearing out the value in `newQueryPlan`. Seems a bit safer to write in the map only from one "place" plus it saves some [very very minor I admit] cost of removing and re-adding an entry in the map